packages/teleterm/src/ui/DocumentCluster/ClusterResources/useServerSideResources.ts

```diff
@@ -0,0 +1,172 @@
+import { useState, useEffect, useMemo } from 'react';
+import { SortType } from 'design/DataTable/types';
+import { useAsync } from 'shared/hooks/useAsync';
+import { AgentFilter, AgentLabel } from 'teleport/services/agents';
+
+import { ServerSideParams } from 'teleterm/services/tshd/types';
+import { useAppContext } from 'teleterm/ui/appContextProvider';
+import { retryWithRelogin } from 'teleterm/ui/utils';
+
+import { useClusterContext } from '../clusterContext';
+
+function addAgentLabelToQuery(filter: AgentFilter, label: AgentLabel) {
+  const queryParts = [];
+
+  // Add existing query
+  if (filter.query) {
+    queryParts.push(filter.query);
+  }
+
+  // If there is an existing simple search,
+  // convert it to predicate language and add it
+  if (filter.search) {
+    queryParts.push(`search("${filter.search}")`);
+  }
+
+  // Create the label query.
+  queryParts.push(`labels["${label.name}"] == "${label.value}"`);
+
+  return queryParts.join(' && ');
+}
+
+const limit = 15;
+
+export function useServerSideResources<Agent>(
+  defaultSort: SortType,
+  fetchFunction: (params: ServerSideParams) => Promise<FetchResponse<Agent>>
+) {
+  const ctx = useAppContext();
+  const { clusterUri, documentUri } = useClusterContext();
+  const [pageIndex, setPageIndex] = useState(0);
+  const [keys, setKeys] = useState<string[]>([]);
+  const [agentFilter, setAgentFilter] = useState<AgentFilter>({
+    sort: defaultSort,
+  });
+
+  // startKey is used here as a way to paginate through agents returned from
+  // their respective rpcs.
+  const [fetchAttempt, fetch] = useAsync(
+    (startKey: string, filter: AgentFilter) =>
+      retryWithRelogin(ctx, documentUri, () =>
+        fetchFunction({
+          ...filter,
+          limit,
+          clusterUri,
+          startKey,
+        })
+      )
+  );
+
+  // If there is no startKey at the current page's index, there is no more data to get
+  function nextPage() {
+    const proposedKey = keys[pageIndex];
+    if (proposedKey) {
+      return () => {
+        setPageIndex(pageIndex + 1);
+      };
+    }
+    return null;
+  }
+
+  // If we are at the first page (index 0), we cannot fetch more previous data
+  function prevPage() {
+    const newPageIndex = pageIndex - 1;
+    if (newPageIndex > -1) {
+      return () => {
+        setPageIndex(newPageIndex);
+      };
+    }
+    return null;
+  }
+
+  useEffect(() => {
```

Comment:
I think this hook needs to provide a cleanup function too. While we're waiting for the results for servers, the user can switch to databases and thus unmount the servers component which will attempt to update its state after the response comes in.

When this gets triggered, I see this warning in the console:

> react-dom.development.js?a814:88 Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
>     in Container (created by ClusterResources)

gravitational/webapps#1377 needs to be merged in first though.